include/dlaf/eigensolver/tridiag_solver/impl.h

```diff
@@ -214,4 +215,109 @@ void TridiagSolver<backend, device, T>::call(Matrix<T, device>& tridiag, Matrix<
                                evecs.readwrite_sender(tile_wrt_local));
   }
 }
+
+// Solve for each tridiagonal tile of @p mat_trd with stedc and save the result in the corresponding
+// diagonal tile of @p mat_ev
+//
+// @p mat_trd is a local matrix of size (n, 2)
+// @p mat_ev is a distributed matrix of size (n, n)
+template <class T, Device D>
+void solveDistLeaf(comm::CommunicatorGrid grid, common::Pipeline<comm::Communicator>& full_task_chain,
+                   Matrix<T, D>& mat_trd, Matrix<T, D>& mat_ev) {
+  const matrix::Distribution& dist = mat_ev.distribution();
+  namespace ex = pika::execution::experimental;
+
+  comm::Index2D this_rank = dist.rankIndex();
+  SizeType ntiles = dist.nrTiles().rows();
+  for (SizeType i = 0; i < ntiles; ++i) {
+    GlobalTileIndex ii_tile(i, i);
+    comm::Index2D ii_rank = dist.rankGlobalTile(ii_tile);
+    GlobalTileIndex idx_trd(i, 0);
+    if (ii_rank == this_rank) {
+      stedcAsync<D>(mat_trd.readwrite_sender(idx_trd), mat_ev.readwrite_sender(ii_tile));
+      ex::start_detached(comm::scheduleSendBcast(full_task_chain(), mat_trd.read_sender(idx_trd)));
+    }
+    else {
+      comm::IndexT_MPI root_rank = grid.rankFullCommunicator(ii_rank);
+      ex::start_detached(
+          comm::scheduleRecvBcast(full_task_chain(), root_rank, mat_trd.readwrite_sender(idx_trd)));
+    }
+  }
+}
+
+// Distributed tridiagonal solver on CPUs
+template <class T>
+void tridiagSolverOnCPU(comm::CommunicatorGrid grid, Matrix<T, Device::CPU>& tridiag,
+                        Matrix<T, Device::CPU>& evals, Matrix<T, Device::CPU>& evecs) {
+  constexpr Device D = Device::CPU;
+
+  // Auxiliary matrix used for the D&C algorithm
+  const matrix::Distribution& dist_evecs = evecs.distribution();
+  const matrix::Distribution& dist_evals = evals.distribution();
+
+  WorkSpace<T, D> ws{Matrix<T, D>(dist_evecs),         // mat1
+                     Matrix<T, D>(dist_evecs),         // mat2
+                     Matrix<T, D>(dist_evals),         // dtmp
+                     Matrix<T, D>(dist_evals),         // z
+                     Matrix<T, D>(dist_evals),         // ztmp
+                     Matrix<SizeType, D>(dist_evals),  // i1
+                     Matrix<SizeType, D>(dist_evals),  // i2
+                     Matrix<SizeType, D>(dist_evals),  // i3
+                     Matrix<ColType, D>(dist_evals)};  // c
+
+  // Set `evecs` to `zero` (needed for Given's rotation to make sure no random values are picked up)
+  matrix::util::set0<Backend::MC, T, D>(pika::execution::thread_priority::normal, evecs);
+
+  // Cuppen's decomposition
+  std::vector<pika::shared_future<T>> offdiag_vals = cuppensDecomposition(tridiag);
+
+  common::Pipeline<comm::Communicator> full_task_chain(grid.fullCommunicator());
+  common::Pipeline<comm::Communicator> row_task_chain(grid.rowCommunicator());
+  common::Pipeline<comm::Communicator> col_task_chain(grid.colCommunicator());
```

Comment:
I think `clone` is missing here too...

Reply:
https://github.com/eth-cscs/DLA-Future/pull/719/commits/0189f92a49066f3374d6a1c0f6580549e87e50f5